storage/innobase/handler/ha_innodb.cc

```diff
@@ -17080,6 +17080,52 @@ int innobase_rollback_by_xid(handlerton* hton, XID* xid)
 	}
 }
 
+/**
+This function is used to rollback one X/Open XA distributed transaction
+which is in the prepared state asynchronously.
+
+It only set the transaction's status to ACTIVE and persist the status.
+The transaction will be rolled back by background rollback thread.
+
+@param[in] hton InnoDB handlerton
+@param[in] xid X/Open XA transaction identification
+
+@return 0 or error number */
+int innobase_recover_rollback_by_xid(handlerton *hton, XID *xid)
+{
+	DBUG_ASSERT(hton == innodb_hton_ptr);
+	DBUG_EXECUTE_IF("innobase_xa_fail", return XAER_RMFAIL;);
+	DBUG_EXECUTE_IF("crash_before_async_binlog_recovery", DBUG_SUICIDE(););
+
+	if (high_level_read_only) {
+		return(XAER_RMFAIL);
+	}
+
+	trx_t *trx = trx_get_trx_by_xid(xid);
+	if (trx == NULL) {
+		return(XAER_RMFAIL);
+	}
+
+  // ddl should not be rolled back through recovery
+  ut_ad(!trx->dict_operation);
+
+#ifdef WITH_WSREP
+	/* If a wsrep transaction is being rolled back during
+	the recovery, we must clear the xid in order to avoid
+	writing serialisation history for rolled back transaction. */
+	if (wsrep_is_wsrep_xid(&trx->xid)) {
+		trx->xid.null();
+	}
+#endif /* WITH_WSREP */
+
+	trx_rollback_for_mysql(trx, true /* fast_rollback */);
+
+	/* Change the trx state to TRX_STATE_ACTIVE. Background rollback thread
+	will rollback this trx later. */
+	trx->state = TRX_STATE_ACTIVE;
```

Comment:
This new state change is not documented in `trx0trx.h`.

I think that it would be good to mention in a comment that only one call of `trx_get_trx_by_xid()` for any given XID can find a transaction object. Subsequent calls by other threads would return `nullptr`. That is what guarantees that no other thread can be modifying the state of the transaction at this point.

What is preventing a race condition between `trx_rollback_all_recovered()` and this assignment? What guarantees that each of these transactions will actually be rolled back? Wouldn’t it be safer to leave the `trx->state` as is, and instead add the transaction to a list that would be processed later? I would suggest to avoid changing `trx_rollback_for_mysql()` at all, and simply invoking `trx_undo_set_state_at_prepare()` here.

Reply:
addressed.